Allow for unknown PNG chunks after image data

diff --git a/Tests/test_file_png.py b/Tests/test_file_png.py
index c94f8eaad..840174444 100644
--- a/Tests/test_file_png.py
+++ b/Tests/test_file_png.py
@@ -596,6 +596,7 @@ def test_apng(self):
         im = Image.open("Tests/images/iss634.apng")
         self.assertEqual(im.get_format_mimetype(), 'image/apng')
 
+        # This also tests reading unknown PNG chunks (fcTL and fdAT) in load_end
         expected = Image.open("Tests/images/iss634.webp")
         self.assert_image_similar(im, expected, 0.23)
 
diff --git a/src/PIL/PngImagePlugin.py b/src/PIL/PngImagePlugin.py
index f3a2eaf21..0669ab216 100644
--- a/src/PIL/PngImagePlugin.py
+++ b/src/PIL/PngImagePlugin.py
@@ -533,14 +533,6 @@ def chunk_acTL(self, pos, length):
         self.im_custom_mimetype = 'image/apng'
         return s
 
-    def chunk_fcTL(self, pos, length):
-        s = ImageFile._safe_read(self.fp, length)
-        return s
-
-    def chunk_fdAT(self, pos, length):
-        s = ImageFile._safe_read(self.fp, length)
-        return s
-
 
 # --------------------------------------------------------------------
 # PNG reader
@@ -682,6 +674,9 @@ def load_end(self):
                 break
             except EOFError:
                 ImageFile._safe_read(self.fp, length)
+            except AttributeError:
+                logger.debug("%r %s %s (unknown)", cid, pos, length)
+                ImageFile._safe_read(self.fp, length)
         self._text = self.png.im_text
         self.png.close()
         self.png = None
